File: app/src/main/java/io/stamethyst/DisplayConfigSync.java
package io.stamethyst;

import android.content.Context;

import java.io.File;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

final class DisplayConfigSync {
    private static final int DEFAULT_FPS_LIMIT = 60;
    private static final int FALLBACK_TARGET_FPS_LIMIT = 120;
    private static final boolean DEFAULT_FULLSCREEN = false;
    private static final boolean DEFAULT_WINDOWED_FULLSCREEN = false;
    private static final boolean DEFAULT_VSYNC = true;
    private static final int MIN_WIDTH = 800;
    private static final int MIN_HEIGHT = 450;

    private DisplayConfigSync() {
    }

    static void syncToCurrentResolution(Context context, int width, int height, int targetFpsLimit) throws IOException {
        int safeWidth = Math.max(MIN_WIDTH, width);
        int safeHeight = Math.max(MIN_HEIGHT, height);
        int normalizedTargetFpsLimit = normalizeTargetFpsLimit(targetFpsLimit);

        File configFile = RuntimePaths.displayConfigFile(context);
        DisplayConfigState state = readExisting(configFile).withFpsLimit(normalizedTargetFpsLimit);
        writeConfig(configFile, safeWidth, safeHeight, state);
    }

    private static DisplayConfigState readExisting(File configFile) {
        if (!configFile.isFile()) {
            return DisplayConfigState.defaults();
        }
        List<String> lines;
        try {
            lines = Files.readAllLines(configFile.toPath(), StandardCharsets.UTF_8);
        } catch (Throwable ignored) {
            return DisplayConfigState.defaults();
        }

        int fps = lines.size() > 2 ? parsePositiveInt(lines.get(2), DEFAULT_FPS_LIMIT) : DEFAULT_FPS_LIMIT;
        boolean fullscreen = lines.size() > 3 ? parseBoolean(lines.get(3), DEFAULT_FULLSCREEN) : DEFAULT_FULLSCREEN;
        boolean wfs = lines.size() > 4 ? parseBoolean(lines.get(4), DEFAULT_WINDOWED_FULLSCREEN) : DEFAULT_WINDOWED_FULLSCREEN;
        boolean vsync = lines.size() > 5 ? parseBoolean(lines.get(5), DEFAULT_VSYNC) : DEFAULT_VSYNC;
        return new DisplayConfigState(fps, fullscreen, wfs, vsync);
    }

    private static void writeConfig(File configFile, int width, int height, DisplayConfigState state) throws IOException {
        File parent = configFile.getParentFile();
        if (parent != null && !parent.exists() && !parent.mkdirs()) {
            throw new IOException("Failed to create directory: " + parent.getAbsolutePath());
        }

        List<String> lines = new ArrayList<>(6);
        lines.add(Integer.toString(width));
        lines.add(Integer.toString(height));
        lines.add(Integer.toString(state.fpsLimit));
        lines.add(Boolean.toString(state.fullscreen));
        lines.add(Boolean.toString(state.windowedFullscreen));
        lines.add(Boolean.toString(state.vsync));
        Files.write(configFile.toPath(), lines, StandardCharsets.UTF_8);
    }

    private static int parsePositiveInt(String raw, int fallback) {
        if (raw == null) {
            return fallback;
        }
        try {
            int value = Integer.parseInt(raw.trim());
            return value > 0 ? value : fallback;
        } catch (Throwable ignored) {
            return fallback;
        }
    }

    private static boolean parseBoolean(String raw, boolean fallback) {
        if (raw == null) {
            return fallback;
        }
        String normalized = raw.trim().toLowerCase(Locale.US);
        if ("true".equals(normalized)) {
            return true;
        }
        if ("false".equals(normalized)) {
            return false;
        }
        return fallback;
    }

    private static int normalizeTargetFpsLimit(int targetFpsLimit) {
        if (targetFpsLimit == 60
                || targetFpsLimit == 90
                || targetFpsLimit == 120
                || targetFpsLimit == 240) {
            return targetFpsLimit;
        }
        return FALLBACK_TARGET_FPS_LIMIT;
    }

    private static final class DisplayConfigState {
        private final int fpsLimit;
        private final boolean fullscreen;
        private final boolean windowedFullscreen;
        private final boolean vsync;

        private DisplayConfigState(int fpsLimit, boolean fullscreen, boolean windowedFullscreen, boolean vsync) {
            this.fpsLimit = fpsLimit;
            this.fullscreen = fullscreen;
            this.windowedFullscreen = windowedFullscreen;
            this.vsync = vsync;
        }

        private DisplayConfigState withFpsLimit(int fpsLimit) {
            return new DisplayConfigState(
                    fpsLimit,
                    fullscreen,
                    windowedFullscreen,
                    vsync
            );
        }

        private static DisplayConfigState defaults() {
            return new DisplayConfigState(
                    DEFAULT_FPS_LIMIT,
                    DEFAULT_FULLSCREEN,
                    DEFAULT_WINDOWED_FULLSCREEN,
                    DEFAULT_VSYNC
            );
        }
    }
}

File: app/src/main/java/io/stamethyst/backend/render/DisplayConfigSync.kt
package io.stamethyst.backend.render

import android.content.Context
import io.stamethyst.config.RuntimePaths
import java.io.File
import java.io.IOException
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.util.Locale

object DisplayConfigSync {
    private const val DEFAULT_WIDTH = 1280
    private const val DEFAULT_HEIGHT = 720
    private const val DEFAULT_FPS_LIMIT = 60
    private const val DEFAULT_FULLSCREEN = false
    private const val DEFAULT_WINDOWED_FULLSCREEN = false
    private const val DEFAULT_VSYNC = true
    private const val MIN_WIDTH = 1
    private const val MIN_HEIGHT = 1
    private val SUPPORTED_FPS_LIMITS = intArrayOf(24, 30, 60, 120, 240)

    @JvmStatic
    @Throws(IOException::class)
    fun syncToCurrentResolution(
        context: Context,
        width: Int,
        height: Int,
        targetFpsLimitOverride: Int? = null
    ) {
        val configFile = RuntimePaths.displayConfigFile(context)
        val lines = buildConfigLines(
            existingLines = readExistingLines(configFile),
            width = width,
            height = height,
            targetFpsLimitOverride = targetFpsLimitOverride
        )
        writeConfig(configFile, lines)
    }

    @JvmStatic
    fun readTargetFpsLimit(context: Context): Int {
        val configFile = RuntimePaths.displayConfigFile(context)
        return normalizeTargetFpsLimit(readExisting(readExistingLines(configFile)).fpsLimit)
    }

    @JvmStatic
    @Throws(IOException::class)
    fun saveTargetFpsLimit(context: Context, targetFpsLimit: Int) {
        val configFile = RuntimePaths.displayConfigFile(context)
        val lines = buildTargetFpsConfigLines(
            existingLines = readExistingLines(configFile),
            targetFpsLimit = targetFpsLimit
        )
        writeConfig(configFile, lines)
    }

    internal fun buildConfigLines(
        existingLines: List<String>?,
        width: Int,
        height: Int,
        targetFpsLimitOverride: Int? = null
    ): List<String> {
        // Keep the display config aligned with the actual scaled render surface.
        // If this file clamps to a larger minimum than the active surface/window size,
        // the game can render into only a corner of the visible output.
        val safeWidth = width.coerceAtLeast(MIN_WIDTH)
        val safeHeight = height.coerceAtLeast(MIN_HEIGHT)
        val state = readExisting(existingLines)
        val fpsLimit = if (targetFpsLimitOverride != null) {
            normalizeTargetFpsLimit(targetFpsLimitOverride)
        } else {
            state.fpsLimit
        }

        return ArrayList<String>(6).apply {
            add(safeWidth.toString())
            add(safeHeight.toString())
            add(fpsLimit.toString())
            add(java.lang.Boolean.toString(state.fullscreen))
            add(java.lang.Boolean.toString(state.windowedFullscreen))
            add(java.lang.Boolean.toString(state.vsync))
        }
    }

    internal fun buildTargetFpsConfigLines(
        existingLines: List<String>?,
        targetFpsLimit: Int
    ): List<String> {
        val normalizedTargetFpsLimit = normalizeTargetFpsLimit(targetFpsLimit)
        val state = readExisting(existingLines).withFpsLimit(normalizedTargetFpsLimit)
        val width =
            if (existingLines != null && existingLines.isNotEmpty()) {
                parsePositiveInt(existingLines[0], DEFAULT_WIDTH)
            } else {
                DEFAULT_WIDTH
            }
        val height =
            if (existingLines != null && existingLines.size > 1) {
                parsePositiveInt(existingLines[1], DEFAULT_HEIGHT)
            } else {
                DEFAULT_HEIGHT
            }

        return ArrayList<String>(6).apply {
            add(width.toString())
            add(height.toString())
            add(state.fpsLimit.toString())
            add(java.lang.Boolean.toString(state.fullscreen))
            add(java.lang.Boolean.toString(state.windowedFullscreen))
            add(java.lang.Boolean.toString(state.vsync))
        }
    }

    private fun readExistingLines(configFile: File): List<String>? {
        if (!configFile.isFile) {
            return null
        }
        return try {
            Files.readAllLines(configFile.toPath(), StandardCharsets.UTF_8)
        } catch (_: Throwable) {
            null
        }
    }

    private fun readExisting(lines: List<String>?): DisplayConfigState {
        if (lines == null) {
            return DisplayConfigState.defaults()
        }
        val fps =
            if (lines.size > 2) parsePositiveInt(lines[2], DEFAULT_FPS_LIMIT) else DEFAULT_FPS_LIMIT
        val fullscreen =
            if (lines.size > 3) parseBoolean(lines[3], DEFAULT_FULLSCREEN) else DEFAULT_FULLSCREEN
        val wfs =
            if (lines.size > 4) parseBoolean(lines[4], DEFAULT_WINDOWED_FULLSCREEN) else DEFAULT_WINDOWED_FULLSCREEN
        val vsync =
            if (lines.size > 5) parseBoolean(lines[5], DEFAULT_VSYNC) else DEFAULT_VSYNC
        return DisplayConfigState(fps, fullscreen, wfs, vsync)
    }

    @Throws(IOException::class)
    private fun writeConfig(configFile: File, lines: List<String>) {
        val parent = configFile.parentFile
        if (parent != null && !parent.exists() && !parent.mkdirs()) {
            throw IOException("Failed to create directory: ${parent.absolutePath}")
        }
        Files.write(configFile.toPath(), lines, StandardCharsets.UTF_8)
    }

    private fun parsePositiveInt(raw: String?, fallback: Int): Int {
        if (raw == null) {
            return fallback
        }
        return try {
            val value = raw.trim().toInt()
            if (value > 0) value else fallback
        } catch (_: Throwable) {
            fallback
        }
    }

    private fun parseBoolean(raw: String?, fallback: Boolean): Boolean {
        if (raw == null) {
            return fallback
        }
        val normalized = raw.trim().lowercase(Locale.US)
        if ("true" == normalized) {
            return true
        }
        if ("false" == normalized) {
            return false
        }
        return fallback
    }

    private fun normalizeTargetFpsLimit(targetFpsLimit: Int): Int {
        return if (SUPPORTED_FPS_LIMITS.contains(targetFpsLimit)) {
            targetFpsLimit
        } else {
            DEFAULT_FPS_LIMIT
        }
    }

    private data class DisplayConfigState(
        val fpsLimit: Int,
        val fullscreen: Boolean,
        val windowedFullscreen: Boolean,
        val vsync: Boolean
    ) {
        fun withFpsLimit(nextFpsLimit: Int): DisplayConfigState {
            return DisplayConfigState(nextFpsLimit, fullscreen, windowedFullscreen, vsync)
        }

        fun withVsync(nextVsync: Boolean): DisplayConfigState {
            return DisplayConfigState(fpsLimit, fullscreen, windowedFullscreen, nextVsync)
        }

        companion object {
            fun defaults(): DisplayConfigState {
                return DisplayConfigState(
                    DEFAULT_FPS_LIMIT,
                    DEFAULT_FULLSCREEN,
                    DEFAULT_WINDOWED_FULLSCREEN,
                    DEFAULT_VSYNC
                )
            }
        }
    }
}
